### ISSUE TITLE
sphere-into-box test produces no collision when sphere fully inside box

### ISSUE THREAD

**Issue by ArsThaumaturgis** [panda3d/panda3d#1335]
## Description
I'd like to see detection of collisions between CollisionBoxes and other (compatible) collision-solids in cases in which those other solids are entirely contained within the CollisionBoxes.

## Use Case
One advantage provided by the built-in three-dimensional collision-solids is that they have a defined "interior", and that they can thus detect objects that are "within" them.

This can be useful in those cases in which one doesn't want to rely on a collider touching the edges of the solid, whether due to size, speed, or even simply manner of positioning.

(For example, I have a situation in which I want objects to be detected by triggers--even if they're "teleported" into position. At current, that latter case doesn't work with CollisionBoxes.)

## Test Program

The following program illustrates (on my machine at least) the current state of such collisions, and may be useful as a test-case.

In short, it creates a large CollisionBox and a smaller CollisionSphere (along with attendant CollisionNodes and NodePaths, of course), and prints out a piece of text when the two are deemed to be "colliding".

Three lines control the test, all marked by arrows in comments: The first two allow one to swap in a CollisionSphere for the CollisionBox and vice versa, demonstrating the "interior" behaviour of the sphere, while the last allows one to place the smaller sphere such that it intersects a wall of the box.

```python
from direct.showbase.ShowBase import ShowBase

from panda3d.core import Vec4, CollisionSphere, CollisionBox, CollisionHandlerEvent, CollisionTraverser, CollisionNode

import math

from panda3d import __version__ as pandaVersion
print (pandaVersion)

import sys
print (sys.version)

class Game(ShowBase):
    def __init__(self):
        ShowBase.__init__(self)

        self.cTrav = CollisionTraverser()
        self.handler = CollisionHandlerEvent()
        self.handler.addInPattern("%fn-into")
        self.handler.addAgainPattern("%fn-within")

        self.accept("obj-into", self.collisionDetected)
        self.accept("obj-within", self.collisionDetected)

        trigger = CollisionNode("trigger")
        solid = CollisionBox((0, 0, 0), 5, 5, 5)    # <-----
        #solid = CollisionSphere(0, 0, 0, 5)        # <-----
        trigger.addSolid(solid)
        trigger.setIntoCollideMask(1)
        trigger.setFromCollideMask(0)
        triggerNP = render.attachNewNode(trigger)
        triggerNP.setY(32)
        triggerNP.show()

        obj = CollisionNode("obj")
        solid = CollisionSphere(0, 0, 0, 2)
        obj.addSolid(solid)
        obj.setFromCollideMask(1)
        obj.setIntoCollideMask(0)
        objNP = render.attachNewNode(obj)
        self.cTrav.addCollider(objNP, self.handler)
        objNP.setY(30)
        #objNP.setX(5)       # <----
        objNP.show()
        self.objNP = objNP

    def collisionDetected(self, entry):
        print ("Collision detected!")


app = Game()
app.run()
```

**Comment by rdb** [comment]
Sorry, I do not understand at all what you are asking for here.  Are you saying that you wish to be able to know whether a solid is entirely contained within another one, instead of just partially?

**Comment by ArsThaumaturgis** [comment]
My apologies.

In short, I want collisions to be detected (at all) in the case in which a (compatible) solid is entirely within a CollisionBox, as is already the case for (compatible) solids entirely within a CollisionSphere.

Put it this way:

If we have a CollisionSphere of radius 5, and a small CollisionSphere of radius 1, both located at the same position, there is no overlap of their surfaces. However, the collisions system will (if I'm not much mistaken), nevertheless consider this to be a collision, and respond accordingly.

Conversely, however, if we have a CollisionBox of size 5 and a small CollisionSphere of radius 1, both located at the same position, there is gain no overlap of their surfaces--and the collision system will _not_ (if I'm not much mistaken) consider this to be a collision.

I thus want the second case to act in the same manner as the first: for an object entirely within a CollisionBox to be detected as "colliding" with that box.

**Comment by rdb** [comment]
This should already be the case, because boxes are solid. If not, that is a bug. Can you show a test that demonstrates otherwise?

**Comment by ArsThaumaturgis** [comment]
Ah, I see.

Well, on my machine at least the code that I gave in the first screenshot above displays the issue.

**Comment by rdb** [comment]
Yes, bug confirmed for the sphere-into-box test.  It's been written to not consider a sphere as colliding if the sphere is not colliding with any of the sides of the box.  This is an error.

I'm tempted to just replace the test with the box-into-sphere I had already rewritten some time ago, I am pretty sure it is more efficient anyway.  But, then this fix will likely only make it into 1.11.  Is that OK with you, or do you need a fix in 1.10?

**Comment by ArsThaumaturgis** [comment]
Ah, I'm glad that the source of the issue has been found! ^_^

As to where the fix is released, it would be ideal for me to have it sooner--but I daresay that I can make do without.

(It primarily limits the shapes of trigger that I can have in my game-world, but I should be able to work with triggers composed of spheres and capsules, I think.)

**Comment by ArsThaumaturgis** [comment]
A question, if I may: What would a fix in 1.10 entail? You seem to suggest that the fix that you have in mind would only be applied towards 1.11--do you have some other fix in mind that might be applied to 1.10, then?

I ask because I'm finding that there are cases in which a box-collider really is the shape that I want. (For example in placing a trigger that covers an entire, rectangular room.) While replacing these with, say, a series of capsules is feasible, it's also not ideal.

Thus I'm wondering what would be incurred by asking for a fix in 1.10.

**Comment by rdb** [comment]
A fix in 1.10 means I will dig through the existing test code to make it work rather than rewrite the test.

Let's see if I can do that this week, otherwise it'll have to wait until 1.10.13.

**Comment by ArsThaumaturgis** [comment]
Ah, I see--fair enough!

In that case I leave the choice of which is better up to your wisdom--while this feature would be rather useful for me to have for my current work, I would in the long term be inclined to want Panda to have the better test in place.

**Comment by rdb** [comment]
I checked in a fix for 1.10.12.  It now does report a collision even if the sphere is entirely inside the box.  The surface normal and interior point aren't quite right, but I wanted to touch the test only minimally out of fear of breaking something.

**Comment by ArsThaumaturgis** [comment]
Ah, that's wonderful, thank you! :D

(In my case I don't really need the surface normal or interior point in many cases; those things for which I do want surface-normals I tend to have represented by spheres or capsules anyway, I believe.)
